### PR TITLE
Iterative

### PR DESCRIPTION
**Pull request by Sam-Bouten**
Added convenience class FDNumericalData to convert and store numerical arrays
in-between iterations, and so avoid recomputation.

Added iterative solver with initial setup (constant numerical data)
and updated part (solving for coordinates and dependent variables).

Added test script for iterative solver.

### CODE REVIEW

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                  edges: List[Tuple[int, int]],
+                  forcedensities: List[float],
+                  loads: Optional[Union[Sequence[Annotated[List[float], 3]], NDArray[(Any, 3), float64]]] = None,
+                  constraints: Sequence[Constraint],
+                  max_iter: int = 100, tol_res: float = 1E-3, tol_xyz: float = 1E-3
+                  ) -> Result:
+    """Iteratively compute the equilibrium coordinates of a system of vertices connected by edges.
+    Vertex constraints are recomputed at each iteration.
+    """
+    nd = FDNumericalData.from_fd_params(vertices, fixed, edges, forcedensities, loads)
```
[the remark this answers is not in the document]
Comment (brgcode) [reply src/compas_fd/fd/fd_iter_numpy.py:36]: personally i would also prefer `data` over `nd`. `nd` can mean all sorts of things...

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    for k in range(max_iter):
+        _xyz = nd.xyz
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:39]: why the underscore?

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def fd_iter_numpy(*,
+                  vertices: Union[Sequence[Annotated[List[float], 3]], NDArray[(Any, 3), float64]],
+                  fixed: List[int],
+                  edges: List[Tuple[int, int]],
+                  forcedensities: List[float],
+                  loads: Optional[Union[Sequence[Annotated[List[float], 3]], NDArray[(Any, 3), float64]]] = None,
+                  constraints: Sequence[Constraint],
+                  max_iter: int = 100, tol_res: float = 1E-3, tol_xyz: float = 1E-3
```
[the remark this answers is not in the document]
Comment (brgcode) [reply src/compas_fd/fd/fd_iter_numpy.py:31]: also, what is `tol_xyz`? tolerance on the displacements between iterations? perhaps `tol_disp` or `tol_dxyz` would be better.

also, docstrings... :)

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        _update_constraints(nd, constraints)
+        if (_is_converged_res(nd.tr, tol_res) and
+           _is_converged_xyz(_xyz, nd.xyz, tol_xyz)):
+            break
+
+    return nd.to_result()
+
+
+def _solve_fd(num_data: FDNumericalData
+              ) -> None:
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:50]: no need to put this on a separate line 
https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    """Solve a single iteration for the equilibrium coordinates of a system.
+    All updated numerical arrays are stored in the num_data parameter.
+    """
+    free, fixed, xyz, q, Q, p, C, Ai, Af, *_ = astuple(num_data)
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:54]: in an earlier function you used `nd`.
although https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds
some consistency helps :)

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    b = p[free] - Af.dot(xyz[fixed])
+    xyz[free] = spsolve(Ai, b)
+    lengths = normrow(C.dot(xyz))
+    num_data.f = q * lengths
+    num_data.r = p - C.T.dot(Q).dot(C).dot(xyz)
+    num_data.ln = lengths
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:60]: why not just `num_data.lengths = lengths`?
except for you, the author, nobody can guess that this is the correct abbreviation for lengths

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        if not constraint:
+            continue
+        constraint.location = xyz[vertex]
+        constraint.residual = residuals[vertex]
+        xyz[vertex] = constraint.location + constraint.tangent * 0.5
+    num_data.tr = asarray([c.tangent for c in constraints if c])
+    num_data.xyz = xyz
+
+
+def _is_converged_res(residuals: NDArray[(Any, 3), float64],
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:82]: res ... ults? iduals?

File: src/compas_fd/fd/fd_iter_numpy.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def _is_converged_xyz(old_xyz: NDArray[(Any, 3), float64],
+                      new_xyz: NDArray[(Any, 3), float64],
+                      tol_xyz: float
+                      ) -> bool:
+    """Validate whether the maximum coordinate displacement
+    between consecutive iterations is within tolerance.
+    """
+    max_dxyz = max(norm(new_xyz - old_xyz, axis=1))
```
Comment (brgcode) [inline src/compas_fd/fd/fd_iter_numpy.py:100]: `dxyz` indeed

File: src/compas_fd/fd/fd_iter_numpy.py (hunk 5 above)
Comment (brgcode) [reply src/compas_fd/fd/fd_iter_numpy.py:54]: btw, `numdata` would be kinda nice

File: src/compas_fd/fd/fd_numerical_data.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    xyz: NDArray[(Any, 3), float64]
+    q: NDArray[(Any, 1), float64]
+    Q: NDArray[(Any, Any), float64]
+    p: NDArray[(Any, 1), float64]
+    C: NDArray[(Any, Any), int]
+    Ai: NDArray[(Any, Any), float64]
+    Af: NDArray[(Any, Any), float64]
+    f: NDArray[(Any, 1), float64] = None
+    r: NDArray[(Any, 3), float64] = None
+    tr: NDArray[(Any, 3), float64] = None
```
Comment (brgcode) [inline src/compas_fd/fd/fd_numerical_data.py:37]: `tr`?

File: src/compas_fd/fd/fd_numerical_data.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+from numpy import zeros_like
+from scipy.sparse import diags
+
+from compas.numerical import connectivity_matrix
+
+from .result import Result
+
+
+@dataclass
+class FDNumericalData:
```
Comment (brgcode) [inline src/compas_fd/fd/fd_numerical_data.py:23]: i think this thing is a very good initiative, but it is meant to improve readability not make things more cryptic. why not write out (some of) the names of the attributes?

File: src/compas_fd/fd/mesh_fd_iter_numpy.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    Returns
+    -------
+    :class:`compas_fd.datastructures.CableMesh`
+        The function updates the mesh in place,
+        but returns a reference to the updated mesh as well
+        for compatibility with RPCs.
+
+    """
+    k_i = mesh.key_index()
+    result = fd_iter_numpy(vertices=array(mesh.vertices_attributes('xyz'), dtype=float64),
```
Comment (brgcode) [inline src/compas_fd/fd/mesh_fd_iter_numpy.py:25]: again, readability... why not define the values of these parameters first?

File: src/compas_fd/fd/mesh_fd_iter_numpy.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+def _update_mesh(mesh, result):
+    for key, attr in mesh.vertices(True):
+        index = mesh.key_index()[key]
+        attr['x'] = result.vertices[index, 0]
+        attr['y'] = result.vertices[index, 1]
+        attr['z'] = result.vertices[index, 2]
+        attr['_rx'] = result.residuals[index, 0]
+        attr['_ry'] = result.residuals[index, 1]
+        attr['_rz'] = result.residuals[index, 2]
```
Comment (brgcode) [inline src/compas_fd/fd/mesh_fd_iter_numpy.py:47]: written out, this is very digestible...

**Review by brgcode — CHANGES_REQUESTED** [review]

File: src/compas_fd/fd/fd_iter_numpy.py (hunk 2 above)
Comment (Sam-Bouten) [reply src/compas_fd/fd/fd_iter_numpy.py:39]: to differentiate it as the previous iteration xyz,
I will rename it to xyz_prev

File: src/compas_fd/fd/fd_iter_numpy.py (hunk 5 above)
Comment (Sam-Bouten) [reply src/compas_fd/fd/fd_iter_numpy.py:54]: numdata it is, consistently (:

**Pull request merged**